src/main/java/org/rascalmpl/maven/CompileRascalMojo.java

```diff
@@ -256,404 +146,123 @@ private int parallelAmount() {
-			}
-			else {
-				prePhaseDone.release(chunks.size() + 1);
-			}
+			List<Process> processes = new LinkedList<>();
+			Process prechecker = runMain(verbose, chunks.get(0), srcs, libs, tmpGeneratedSources.get(0), tmpBins.get(0));
+			result += prechecker.waitFor(); // block until the process is finished
 
-			// running parallel jobs for the identified chunks
-			if (!chunks.isEmpty()) {
-				safeLog(l -> l.debug("Preparing checker for in " + chunks.size() + " parallel threads"));
-				try {
-					List<ISourceLocation> binFolders = new ArrayList<>();
-					List<ISourceLocation> generateSourcesFolders = new ArrayList<>();
-					Semaphore done = new Semaphore(0);
-					for (IList todo: chunks) {
-						ISourceLocation myBin = VF.sourceLocation("tmp", "","tmp-" + System.identityHashCode(todo) + "-" + Instant.now().getEpochSecond());
-						ISourceLocation mySources = VF.sourceLocation("tmp", "","tmp-srcs-" + System.identityHashCode(todo) + "-" + Instant.now().getEpochSecond());
-						binFolders.add(myBin);
-						generateSourcesFolders.add(mySources);
-						PathConfig myConfig = new PathConfig(pcfg.getSrcs(), pcfg.getLibs().append(pcfg.getBin()), myBin);
-
-						executor.execute(() -> {
-							try {
-								Thread.sleep(1000);  // give the other evaluator a head start
-								safeLog(l -> l.debug("Starting fresh evaluator"));
-								parallelReports.add(evaluators.useAndReturn(e -> {
-									try {
-										prePhaseDone.acquire();
-										safeLog(l -> l.debug("Starting checking chunk with " + todo.size() +  " entries"));
-
-										var epcfg = expandPathConfig(myConfig, mySources);
-										var myCompilerConfig = makeCompilerConfig(e, verbose, epcfg);
-
-										return runCheckerSingle(e.getMonitor(), todo, e, myCompilerConfig);
-									}
-									catch (InterruptedException interruptedException) {
-									    return VF.list();
-									}
-								}));
-							}
-							catch (Exception e) {
-								safeLog(l -> l.error("Failure executing:", e));
-								failure.compareAndSet(null, e);
-							}
-							finally {
-								done.release();
-							}
-						});
-					}
-					done.acquire(chunks.size());
-
-					// now have to merge the result bins (output files of each build)
-					// it's possible single modules are produced by different chunks
-					// but they are guaranteed to be the same
-					mergeOutputFolders(pcfg.getBin(), binFolders, false);
-					// we also copy of the generated sources, but it's fine right now if they are ignored
-					mergeOutputFolders(generatedSourcesLoc, generateSourcesFolders, true);
-
-				} catch (URISyntaxException | IOException | InterruptedException e) {
-					getLog().error("Failed post-processing evaluator", e);
-					failure.compareAndSet(null, e);
-				}
+			// starts the processes asynchronously
+			for (int i = 1; i < chunks.size(); i++) {
+				processes.add(runMain(verbose, chunks.get(1), srcs, libs, tmpGeneratedSources.get(i), tmpBins.get(i)));
```

Comment:
Shouldn't this be `chunks.get(1)` but `i`?

And also, we should put the bin folder of the preCheck phase on the lib path of the parallel runner. so that it doesn't have to start all over again and recheck `List` etc.

Reply:
also: since runMain uses `inheritIO` this will start "biting" each other.

It would be better to have the process allocate it's own streams, and then copy from them in a streaming way, making sure to lock before we write to our own streams. (like was happening in the original code, but it was doing this inside of the same JVM)

Reply:
yes, with the downside that while the typecheck is happening, you have no feedback. so it looks like it's stuck.

Reply:
> And also, we should put the bin folder of the preCheck phase on the lib path of the parallel runner. so that it doesn't have to start all over again and recheck List etc

nice catch!